tracer/src/Datadog.Trace/DiagnosticListeners/AspNetCoreDiagnosticObserver.cs

```diff
@@ -527,16 +560,41 @@ private void OnRoutingEndpointMatched(object arg)
                 RouteValueDictionary routeValues = request.RouteValues;
                 // No need to ToLowerInvariant() these strings, as we lower case
                 // the whole route later
-                object raw;
-                string controllerName = routeValues.TryGetValue("controller", out raw)
-                                            ? raw as string
-                                            : null;
-                string actionName = routeValues.TryGetValue("action", out raw)
-                                        ? raw as string
-                                        : null;
-                string areaName = routeValues.TryGetValue("area", out raw)
-                                      ? raw as string
-                                      : null;
+                string controllerName = null;
+                string actionName = null;
+                string areaName = null;
+
+                // Let's avoid using TryGetValue multiple times and enumerate the Dictionary and bailout when we get all the data.
+                var finishBits = 0;
+                foreach (var routeValue in routeValues)
+                {
+                    // 1 | 2 | 4 Flags
+                    if (finishBits == 7)
+                    {
+                        break;
+                    }
+
+                    if (routeValue.Key == "controller")
```

Comment:
We need to do case invariant checks here

Reply:
done in https://github.com/DataDog/dd-trace-dotnet/pull/4390/commits/650371eb9f01ba9d28929c87761a48f80b46d8d0